### ISSUE TITLE
Feature Request: can we make night mode ignore photos from being inverted

### ISSUE THREAD

**Issue by fei-hao** [koreader/koreader#2571]
* KOReader version:
* Device:kobo aura 

**It would be great to ignore them from being inverted.**

**Thanks**

![a78c0563a026e149af9a6c012711c7b2](https://cloud.githubusercontent.com/assets/9626442/23293073/f5fb6446-fa9e-11e6-8cd0-d921c08bd9b6.png)
![d031b5ba0748c68369f0a0dde1a1206b](https://cloud.githubusercontent.com/assets/9626442/23293076/fd2e593a-fa9e-11e6-8b57-4754d430999d.png)

**Comment by houqp** [comment]
This is a non-trivia feature to implement for all document types. However, we can probably achieve this with a special CSS setting for crengine.

**Comment by cges30901** [comment]
Hello, I am a new user of koreader on Android and I have the same problem when reading epub files. So I come up with a workaroud: I created koreader/styletweaks/epub-night.css, and set background color to black and text color to white (without enabling night mode).
```
body {
    color: white;
    background-color: black;
}
```
This is not perfect, because margin is still white. I also tried inverting images and applied night mode, but it didn't work. I don't know how to ignore for other formats.

**Comment by Frenzie** [comment]
Is the margin also white if you stick the black background color on `html` instead?

What do you mean by tried inverting images? You mean like `filter: invert(1)`? That's currently not supported I'm afraid. @poire-z Any idea how hard that'd be to add btw?

**Comment by poire-z** [comment]
No idea. Don't really want to get involved in CSS filters and image manipulation :)

Might be easier to just fix the original issue (so no need for CSS) by passing a flag indicating we are in nighmode to crengine, so it can itself invert images (only images? or colors too?), so when they are later inverted by our frontend nighmode code, they get normal again.
Pinging @NiLuJe who has just put his fingers into the crengine images code :) If that would be possible, we would just need an additional global variable set from frontend.

**Comment by Frenzie** [comment]
Nightmode is

html filter(1)
img filter(1)

to revert the global one.

**Comment by cges30901** [comment]
> Is the margin also white if you stick the black background color on `html` instead?

This has no effect. I guess it is not supported.

> You mean like `filter: invert(1)`?

Yes.

**Comment by NiLuJe** [comment]
c.f., https://github.com/koreader/koreader/issues/1375#issuecomment-423413385 for my thoughts on the subject ;).

They haven't changed. Still have subzero interest in working on this ;).

Could be "easy" enough to check in ImageWidget on our end for the UI, and possiblyLVImageScaledDrawCallback in CRe.

**Comment by Frenzie** [comment]
Same. I was just thinking aloud that theoretically it seems that some of the `filter` properties (like `invert`) might be relatively easy to apply. Emphasis on *might*. ;-)

**Comment by poire-z** [comment]
> Could be "easy" enough to check in ImageWidget on our end for the UI, and possibly LVImageScaledDrawCallback in CRe.

I got that working. I'll PR it once all thr RGB32 stuff is bumped :)
For cre, it was just a matter of setting a flag from frontend in base on the drawBuf and use:
```diff
@@ -1301,6 +1304,8 @@ void LVColorDrawBuf::Draw( LVImageSourceRef img, int x, int y, int width, int he
     //fprintf( stderr, "LVColorDrawBuf::Draw( img(%d, %d), %d, %d, %d, %d\n", img->GetWidth(), img->GetHeight(), x, y, width, height );
     LVImageScaledDrawCallback drawcb( this, img, x, y, width, height, dither );
     img->Decode( &drawcb );
+    if ( _invertImages )
+        InvertRect(x, y, x+width, y+height);
     _drawnImagesCount++;
     _drawnImagesSurface += width*height;
 }
```
and implement `LVColorDrawBuf::InvertRect()`. (Not the most efficient, but quick and not painful to implement, so that should do for night mode).

@NiLuJe : can you confirm if this is the proper way of inverting these 16 or 32bpp (it works, just want to be sure it's fine, mostly with alpha):
```c
 void LVColorDrawBuf::InvertRect(int x0, int y0, int x1, int y1)
 {
[...]
+    if (_bpp==16) {
+        for (int y=y0; y<y1; y++) {
+            lUInt16 * line = (lUInt16 *)GetScanLine(y);
+            for (int x=x0; x<x1; x++) {
+                line[x] = ~line[x];
+            }
+        }
+    }
+    else {
+        for (int y=y0; y<y1; y++) {
+            lUInt32 * line = (lUInt32 *)GetScanLine(y);
+            for (int x=x0; x<x1; x++) {
+                line[x] = ( line[x] & 0xFF000000 ) | ~( line[x] & 0xFFFFFF );
+            }
+        }
+    }
 }
```

For ImageWidget, I was fine with just:
```diff
--- a/frontend/ui/widget/imagewidget.lua
+++ b/frontend/ui/widget/imagewidget.lua
@@ -373,6 +373,9 @@ function ImageWidget:paintTo(bb, x, y)
     if self.dim then
         bb:dimRect(x, y, size.w, size.h)
     end
+    if G_reader_settings:isTrue("night_mode") then
+        bb:invertRect(x, y, size.w, size.h)
+    end
 end
```
for covers and ImageViewer, but with that, our icons and chevrons have a white background in nightmode.
Is it an alpha thing? (Haven't spend much time on that because I'm sure you have the answer :)

**Comment by NiLuJe** [comment]
@poire-z : I'm partial to AND instead of NOT, because right now, NOT behaves mostly okay solely because we're doing unsigned maths, so the wraparound is a defined behavior (i.e., -1 => 255).

TL;DR: `^= 0xFFFF` for RGB16 (invert all 16 bits, there's no alpha), `^= 0x00FFFFFF` for RGB32 (invert 3 color components, leave the alpha alone).

**Comment by NiLuJe** [comment]
As for the icons thingy, on which target BB is that? OTOH, I can't see anything obviously wrong with that, but they *are* currently broken, nightmode or not, @ 32bpp w/ Lua blitting (the fix is in the blitting PR, https://github.com/koreader/koreader-base/pull/878/commits/a9a96cb111f91e1aceaf39b40450813b902713d8)

**Comment by poire-z** [comment]
> TL;DR: ^= 0xFFFF for RGB16 (invert all 16 bits, there's no alpha), ^= 0x00FFFFFF for RGB32 (invert 3 color components, leave the alpha alone).

OK, and obvious, thanks :)

> As for the icons thingy, on which target BB is that?

That was on the emulator, without your base PR yet. And the same with a rebuilt now that your PR is merged.
But I guess we should just not pre-invert icons, as they are fine when just inverted by nightmode.
Dunno if there are exceptions, but just not doing that if they come from a file seems to be fine:
```
    if not self.file and G_reader_settings:isTrue("night_mode") then
        bb:invertRect(x, y, size.w, size.h)
    end
```
So, not for icons from files, But yes with covers, full screen covers, embedded images from cre... anything else I forgot?

**Comment by NiLuJe** [comment]
@poire-z : Not that I can think of, but then I don't use night mode ;).

**Comment by poire-z** [comment]
@NiLuJe : so, I've been testing this. All seems fine with crengine (emulator 32bpp, kobo 8bpp, using XOR :).

But with the ImageViewer/ImageWidget, and this at the end of `ImageWidget:paintTo()`:
```lua
    -- If in night mode, invert all rendered images, so the original is
    -- displayed when the whole screen is inverted by night mode.
    -- Except for our black & white icon files, that we want inverted
    -- in night mode.
    if not self.file and G_reader_settings:isTrue("night_mode") then
        bb:invertRect(x, y, size.w, size.h)
    end
```
It seems all fine on the 32bpp/cblitbuffer emulator.
But it fails doing things well on my Kobo 8bpp or 16bpp. Most photos are fine, but some maps don't get really inverted-inverted.
For example, most maps in https://fr.wikipedia.org/wiki/Saint-Empire_romain_germanique saved as EPUB, like [this one](https://upload.wikimedia.org/wikipedia/commons/thumb/7/74/Map_Thirty_Years_War-fr.svg/220px-Map_Thirty_Years_War-fr.svg.png) : the bottom panel is white (so, good) in night mode, while France is black (while they are both white in night mode on the 32bpp emulator).
Bug in the 8bpp invert, or 8bpp/alpha limitation ?

**Comment by NiLuJe** [comment]
@poire-z : That's probably the few cases when alphablitFrom is called instead of blitFrom, yeah (at least for that example, the bottom panel is transparent, yeah).

That might be a case of the actual BG color being different between C and Lua BB, because the C bb doesn't actually ever honor the invert bit of the BB, it just has a shadow buffer that fully inverted.

Can you check how it looks on 32bpp with the Lua bb?

**Comment by NiLuJe** [comment]
The actual difference being that, for truly transparent pixels, *nothing* is plotted, so, whatever was behind stays as-is. That might be a background paintRect, or possibly 0s for a fresh BB?

**Comment by poire-z** [comment]
> Can you check how it looks on 32bpp with the Lua blitter?

Yep. It's different with the lua blitter on 32bpp emulator: the same (bad) as it is on 8bpp kobo :)

**Comment by NiLuJe** [comment]
I couldn't quite figure out if there was a background being painted behind those !file ImageWidgets, but I'd start looking for that, as that *should* be the color bleeding through.

Or that's a quirk of the C blitter ignoring :invert in all non-invertRect codepaths, which might take some brain-twisting to untangle ;).

**Comment by NiLuJe** [comment]
Incidentally, the same reasoning *probably* explains the bg color issues you initially saw when inverting the icons ;).
